### PR TITLE
BigAutoField as new default in 2.0

### PR DESCRIPTION
**Pull request by kennethreitz**
This pull request uses bigints (`BigAutoField`) as the new default for auto created primary keys in Django applications. 

- New applications will use `BigAutoField` by default, instead of `AutoField`. 
- `SmallAutoField` is introduced, and `AutoField` is aliased to it, for later deprecation. 
- Migration path is to warn about the expensive migration, and guide the user to hardcode their models to the new `SmallAutoField`, if desired (and to the release notes). 

Discussed at length with @andrewgodwin, and paired with him on this. 

Tests are all passing locally.

### CODE REVIEW

File: django/db/migrations/questioner.py [hunk 1]
```diff
@@ -188,6 +192,24 @@ def ask_rename(self, model_name, old_name, new_name, field_instance):
...
+
+    def ask_small_to_big_autofield(self, field_name, model_name):
+        """Adding an auto_now_add field to a model."""
+        if not self.dry_run:
+            choice = self._choice_input(
+                "You have upgraded Django, which will change the field {}.{} to a BigAutoField'. "
+                "The database needs to upgrade from int to bigint, which may take a while, if you have a lot of rows ".format(model_name, field_name),
+                [
+                    "Create upgrade migration (may take a while to apply)",
+                    "Quit, and let me specifity SmallAutoField in models.py",
```
Comment (tomviner) [inline django/db/migrations/questioner.py]: _specify_?

File: django/db/migrations/questioner.py [hunk 2]
```diff
@@ -78,6 +78,10 @@ def ask_auto_now_add_addition(self, field_name, model_name):
         # None means quit
         return None
 
+    def ask_small_to_big_autofield(self, field_name, model_name):
+        """Upgrading a AutoField to a BigAutoField on a model."""
```
Comment (tomviner) [inline django/db/migrations/questioner.py]: _**an** AutoField_?

File: django/db/backends/base/features.py [hunk 3]
```diff
@@ -247,6 +247,9 @@ class BaseDatabaseFeatures:
     # Does the backend support keyword parameters for cursor.callproc()?
     supports_callproc_kwargs = False
 
+    # Does the backend just alias bigint to integter.
```
Comment (tomviner) [inline django/db/backends/base/features.py]: _integer_?

File: django/db/migrations/questioner.py (hunk 2 above)
Comment (kennethreitz) [inline django/db/migrations/questioner.py]: @tomviner thanks for the catches!

**Comment by kennethreitz** [comment]
@tomviner fixed :)

File: django/db/migrations/autodetector.py [hunk 4]
```diff
@@ -916,6 +917,11 @@ def generate_altered_fields(self):
                             preserve_default=preserve_default,
                         )
                     )
+
+                elif isinstance(old_field, models.SmallAutoField) and isinstance(new_field, models.BigAutoField):
+                    if not self.questioner.ask_small_to_big_autofield(field_name, model_name):
+                        raise RuntimeError("Automatic upgrade to BigAutoField refused. \nIf you do not want to upgrade, please define {0} = models.SmallAutoField in your model definition.\n\nPlease see the Django release notes for more information.".format(field_name))
```
Comment (shaib) [inline django/db/migrations/autodetector.py]: `models.SmallAutoField`**()**

File: django/db/migrations/questioner.py [hunk 5]
```diff
@@ -188,6 +192,24 @@ def ask_rename(self, model_name, old_name, new_name, field_instance):
         return self._boolean_input(msg % (model_name, old_name, model_name, new_name,
                                           field_instance.__class__.__name__), False)
 
+
+    def ask_small_to_big_autofield(self, field_name, model_name):
+        """Adding an auto_now_add field to a model."""
```
Comment (shaib) [inline django/db/migrations/questioner.py]: Huh?

File: django/db/migrations/questioner.py [hunk 6]
```diff
@@ -188,6 +192,24 @@ def ask_rename(self, model_name, old_name, new_name, field_instance):
...
+        if not self.dry_run:
+            choice = self._choice_input(
+                "You have upgraded Django, which will change the field {}.{} to a BigAutoField'. "
```
Comment (shaib) [inline django/db/migrations/questioner.py]: Please rephrase in a way which will stay valid for the case where the change is made independently of the upgrade.
Something like  `"You are changing the field {}.{} from a SmallAutoField to a BigAutofield (perhaps as a result of upgrading Django)"`

File: django/db/migrations/questioner.py [hunk 7]
```diff
@@ -188,6 +192,24 @@ def ask_rename(self, model_name, old_name, new_name, field_instance):
...
+                "The database needs to upgrade from int to bigint, which may take a while, if you have a lot of rows ".format(model_name, field_name),
+                [
+                    "Create upgrade migration (may take a while to apply)",
+                    "Quit, and let me specify SmallAutoField in models.py",
+                ]
+            )
+            if choice == 2:
+                return False
+            else:
+                return True
```
Comment (shaib) [inline django/db/migrations/questioner.py]: `return choice==1`

File: django/db/models/fields/__init__.py [hunk 8]
```diff
@@ -960,6 +960,10 @@ def rel_db_type(self, connection):
         return BigIntegerField().db_type(connection=connection)
 
 
+# Aliased for backwards compabitbility.
+AutoField = SmallAutoField
```
Comment (shaib) [inline django/db/models/fields/__init__.py]: I think this is better placed right after the SmallAutoField class

File: tests/migrations/test_operations.py [hunk 9]
```diff
@@ -1309,7 +1309,13 @@ def assertIdTypeEqualsFkType():
                     for c in connection.introspection.get_table_description(cursor, "test_alflpkfk_rider")
                     if c.name == "pony_id"
                 ][0]
-            self.assertEqual(id_type, fk_type)
+
+            # Sqlite only allows primary keys of type integer, however, this is allows bigints.
```
Comment (MarkusH) [inline tests/migrations/test_operations.py]: "this is allows bigints" sounds wrong.

File: django/db/migrations/autodetector.py (hunk 4 above)
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: Can you also fix the line length please. Maximum is at 119 chars.

File: django/db/migrations/questioner.py [hunk 10]
```diff
@@ -188,6 +192,24 @@ def ask_rename(self, model_name, old_name, new_name, field_instance):
         return self._boolean_input(msg % (model_name, old_name, model_name, new_name,
                                           field_instance.__class__.__name__), False)
 
+
+    def ask_small_to_big_autofield(self, field_name, model_name):
+        """Adding an auto_now_add field to a model."""
+        if not self.dry_run:
+            choice = self._choice_input(
+                "You have upgraded Django, which will change the field {}.{} to a BigAutoField'. "
+                "The database needs to upgrade from int to bigint, which may take a while, if you have a lot of rows ".format(model_name, field_name),
```
Comment (MarkusH) [inline django/db/migrations/questioner.py]: Please limit the line length to 119 characters.

**Comment by kennethreitz** [comment]
Will update.

File: django/db/migrations/autodetector.py (hunk 4 above)
Comment (kennethreitz) [inline django/db/migrations/autodetector.py]: Already planned! I saw the failing flake8 tests.

Comment (kennethreitz) [inline django/db/migrations/autodetector.py]: @shaib thanks!

File: django/db/migrations/questioner.py (hunk 5 above)
Comment (kennethreitz) [inline django/db/migrations/questioner.py]: @shaib nice catch — hard to focus on details during the sprints — much appreciated.

File: django/db/migrations/questioner.py (hunk 7 above)
Comment (kennethreitz) [inline django/db/migrations/questioner.py]: @shaib i'll make this change, but @andrewgodwin told me to `return True` here.

File: django/db/models/fields/__init__.py (hunk 8 above)
Comment (kennethreitz) [inline django/db/models/fields/__init__.py]: @shaib agreed!

File: tests/migrations/test_operations.py (hunk 9 above)
Comment (kennethreitz) [inline tests/migrations/test_operations.py]: @MarkusH i'm really not good at details when i'm surrounded by people ;)

File: tests/admin_views/tests.py [hunk 11]
```diff
@@ -3678,8 +3678,8 @@ def test_simple_inline(self):
         self.assertEqual(Widget.objects.count(), 1)
         self.assertEqual(Widget.objects.all()[0].name, "Widget 1 Updated")
 
-    def test_explicit_autofield_inline(self):
-        "A model with an explicit autofield primary key can be saved as inlines. Regression for #8093"
+    def test_explicit_BigAutoField_inline(self):
+        "A model with an explicit BigAutoField primary key can be saved as inlines. Regression for #8093"
```
Comment (felixxm) [inline tests/admin_views/tests.py]: Please use docstring and wrap at 79 characters.

File: tests/basic/tests.py [hunk 12]
```diff
@@ -51,7 +51,7 @@ def test_can_create_instance_using_kwargs(self):
         self.assertEqual(a.headline, 'Third article')
         self.assertEqual(a.pub_date, datetime(2005, 7, 30, 0, 0))
 
-    def test_autofields_generate_different_values_for_each_instance(self):
+    def test_BigAutoFields_generate_different_values_for_each_instance(self):
```
Comment (felixxm) [inline tests/basic/tests.py]: Use lowercase.

File: tests/bulk_create/tests.py [hunk 13]
```diff
@@ -103,12 +103,12 @@ def test_non_auto_increment_pk_efficiency(self):
...
     def test_zero_as_autoval(self):
         """
-        Zero as id for AutoField should raise exception in MySQL, because MySQL
+        Zero as id for BigAutoField should raise exception in MySQL, because MySQL
         does not allow zero for automatic primary key.
         """
         valid_country = Country(name='Germany', iso_two_letter='DE')
         invalid_country = Country(id=0, name='Poland', iso_two_letter='PL')
-        msg = 'The database backend does not accept 0 as a value for AutoField.'
+        msg = 'The database backend does not accept 0 as a value for BigAutoField.'
```
Comment (felixxm) [inline tests/bulk_create/tests.py]: Analogous change in the `DatabaseOperations.validate_autopk_value()` should solve MySQL test failure.

File: tests/many_to_one/tests.py [hunk 14]
```diff
@@ -522,7 +522,7 @@ def test_fk_assignment_and_related_object_cache(self):
         self.assertIsNot(c.parent, p)
         self.assertEqual(c.parent, p)
 
-    def test_fk_to_bigautofield(self):
+    def test_fk_to_BigAutoField(self):
```
Comment (felixxm) [inline tests/many_to_one/tests.py]: I think this unrelated change should be reverted.

File: tests/schema/tests.py [hunk 15]
```diff
@@ -1054,23 +1054,23 @@ def test_alter_implicit_id_to_explicit(self):
...
-        new_field = AutoField(primary_key=True)
+        new_field = SmallAutoField(primary_key=True)
         new_field.model = IntegerPK
         new_field.set_attributes_from_name('i')
 
         with connection.schema_editor() as editor:
             editor.alter_field(IntegerPK, old_field, new_field, strict=True)
 
-    def test_alter_int_pk_to_bigautofield_pk(self):
+    def test_alter_int_pk_to_BigAutoField_pk(self):
```
Comment (felixxm) [inline tests/schema/tests.py]: I think this unrelated change should be reverted.

Comment (kennethreitz) [inline tests/schema/tests.py]: agreed

File: tests/many_to_one/tests.py (hunk 14 above)
Comment (kennethreitz) [inline tests/many_to_one/tests.py]: agreed

File: django/db/migrations/questioner.py [hunk 16]
```diff
@@ -78,6 +78,10 @@ def ask_auto_now_add_addition(self, field_name, model_name):
         # None means quit
         return None
 
+    def ask_small_to_big_autofield(self, field_name, model_name):
+        """Upgrading an AutoField to a BigAutoField on a model."""
+        return self.defaults.get("ask_small_to_big_autofield", False)
```
Comment (kennethreitz) [inline django/db/migrations/questioner.py]: @felixxm all other examples in this file are double quotes.

**Comment by kennethreitz** [comment]
@felixxm thanks for the feedback! integrated.

**Comment by kennethreitz** [comment]
@MarkusH will add a test (perhaps with @andrewgodwin's help, if I get stuck) ASAP

File: django/db/backends/base/features.py [hunk 17]
```diff
@@ -247,6 +247,9 @@ class BaseDatabaseFeatures:
     # Does the backend support keyword parameters for cursor.callproc()?
     supports_callproc_kwargs = False
 
+    # Does the backend just alias bigint to integer.
```
Comment (adamchainz) [inline django/db/backends/base/features.py]: stickler point: this end with a question mark

File: django/db/migrations/questioner.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (adamchainz) [reply django/db/migrations/questioner.py]: We're moving to single quotes everywhere

> Use single quotes for strings, or a double quote if the the string contains a single quote. Don’t waste time doing unrelated refactoring of existing code to conform to this style.

https://docs.djangoproject.com/en/1.11/internals/contributing/writing-code/coding-style/

File: django/db/migrations/questioner.py [hunk 18]
```diff
@@ -188,6 +192,24 @@ def ask_rename(self, model_name, old_name, new_name, field_instance):
...
+    def ask_small_to_big_autofield(self, field_name, model_name):
+        """Upgrading an AutoField to a BigAutoField."""
+        if not self.dry_run:
+            choice = self._choice_input(
+                "You are changing the field {}.{} from a SmallAutoField to a BigAutofield "
+                "(perhaps as a result of upgrading Django). The database needs to upgrade from int to bigint, "
+                "which may take a while, if you have a lot of rows ".format(model_name, field_name),
+                [
+                    "Create upgrade migration (may take a while to apply)",
+                    "Quit, and let me specify SmallAutoField in models.py",
```
Comment (adamchainz) [inline django/db/migrations/questioner.py]: Saying `models.py` isn't always accurate, some apps have their models module split up. Perhaps `"...let me specify SmallAutoField in the {} definition".format(model_name)` is better.

File: django/db/migrations/questioner.py [hunk 19]
```diff
@@ -188,6 +192,24 @@ def ask_rename(self, model_name, old_name, new_name, field_instance):
...
+                ]
+            )
+            if choice == 2:
+                return False
+            else:
+                return choice == 1
```
Comment (adamchainz) [inline django/db/migrations/questioner.py]: this whole if/else can be turned into `return (choice == 1)`

File: django/db/migrations/questioner.py [hunk 20]
```diff
@@ -188,6 +192,24 @@ def ask_rename(self, model_name, old_name, new_name, field_instance):
...
+        return None
```
Comment (adamchainz) [inline django/db/migrations/questioner.py]: `False` is a better default here, the result is used in a boolean context only

File: django/db/models/fields/__init__.py [hunk 21]
```diff
@@ -950,7 +950,11 @@ def formfield(self, **kwargs):
         return None
 
 
-class BigAutoField(AutoField):
+# Aliased for backwards compatibility.
+AutoField = SmallAutoField
```
Comment (adamchainz) [inline django/db/models/fields/__init__.py]: are all operations in all pre-existing migrations definitely going to be fine with the class being renamed?

Comment (kennethreitz) [reply django/db/models/fields/__init__.py]: @adamchainz should be, according to @andrewgodwin

File: django/db/migrations/questioner.py (hunk 18 above)
Comment (kennethreitz) [reply django/db/migrations/questioner.py]: @adamchainz can i add you to my fork and you can help me clean this up?

**Comment by matthiask** [comment]
Today might be a good day to revisit this, seeing as Basecamp 3 just went down for hours because they had to convert int->bigint in one of their tables.

https://m.signalvnoise.com/update-on-basecamp-3-being-stuck-in-read-only-as-of-nov-8-9-22am-cst-c41df1a58352

**Comment by kennethreitz** [comment]
Agreed!

File: django/db/models/fields/__init__.py (hunk 21 above)
Comment (charettes) [reply django/db/models/fields/__init__.py]: Pretty sure this will cause problems

https://github.com/django/django/blob/a375e911efcc78054d09ef31422e081507e56623/django/db/models/fields/__init__.py#L475

File: django/db/models/fields/__init__.py [hunk 22]
```diff
@@ -873,7 +873,7 @@ def value_from_object(self, obj):
         return getattr(obj, self.attname)
 
 
-class AutoField(Field):
+class SmallAutoField(Field):
```
Comment (charettes) [inline django/db/models/fields/__init__.py]: This should be kept as-is. We use the `Small` wording for small ints and omit the prefix for 32-bit integers. e.g. `SmallIntegerField`, `IntegerField`, and `BigIntegerField`.

**Comment by auvipy** [comment]
@kenneth-reitz sorry for bugging, could you please retarget this for django 3.0? what is our thought regarding the last observation Simon made? do you need any help to complete this?

File: tests/admin_scripts/app_waiting_migration/migrations/0001_initial.py [hunk 23]
```diff
@@ -12,7 +12,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             name='Bar',
             fields=[
-                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
```
Comment (auvipy) [inline tests/admin_scripts/app_waiting_migration/migrations/0001_initial.py]: this could be reduced to initial migration.

File: tests/migrations/migrations_test_apps/conflicting_app_with_dependencies/migrations/0002_second.py [hunk 24]
```diff
@@ -15,7 +15,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             "Book",
             [
```
Comment (auvipy) [inline tests/migrations/migrations_test_apps/conflicting_app_with_dependencies/migrations/0002_second.py]: this could be removed and regen

File: tests/migrations/migrations_test_apps/lookuperror_a/migrations/0002_a2.py [hunk 25]
```diff
@@ -11,7 +11,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             name='A2',
             fields=[
-                ('id', models.AutoField(verbose_name='ID', primary_key=True, serialize=False, auto_created=True)),
+                ('id', models.BigAutoField(verbose_name='ID', primary_key=True, serialize=False, auto_created=True)),
```
Comment (auvipy) [inline tests/migrations/migrations_test_apps/lookuperror_a/migrations/0002_a2.py]: this could be removed and regen

File: tests/migrations/migrations_test_apps/lookuperror_a/migrations/0004_a4.py [hunk 26]
```diff
@@ -11,7 +11,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             name='A4',
             fields=[
-                ('id', models.AutoField(auto_created=True, serialize=False, verbose_name='ID', primary_key=True)),
+                ('id', models.BigAutoField(auto_created=True, serialize=False, verbose_name='ID', primary_key=True)),
```
Comment (auvipy) [inline tests/migrations/migrations_test_apps/lookuperror_a/migrations/0004_a4.py]: same here

File: tests/migrations/migrations_test_apps/lookuperror_a/migrations/0003_a3.py [hunk 27]
```diff
@@ -13,7 +13,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             name='A3',
             fields=[
-                ('id', models.AutoField(serialize=False, auto_created=True, primary_key=True, verbose_name='ID')),
+                ('id', models.BigAutoField(serialize=False, auto_created=True, primary_key=True, verbose_name='ID')),
```
Comment (auvipy) [inline tests/migrations/migrations_test_apps/lookuperror_a/migrations/0003_a3.py]: remove this plz

File: tests/migrations/migrations_test_apps/lookuperror_b/migrations/0002_b2.py [hunk 28]
```diff
@@ -12,7 +12,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             name='B2',
             fields=[
-                ('id', models.AutoField(primary_key=True, verbose_name='ID', auto_created=True, serialize=False)),
+                ('id', models.BigAutoField(primary_key=True, verbose_name='ID', auto_created=True, serialize=False)),
```
Comment (auvipy) [inline tests/migrations/migrations_test_apps/lookuperror_b/migrations/0002_b2.py]: remove plz

File: tests/migrations/migrations_test_apps/lookuperror_b/migrations/0003_b3.py [hunk 29]
```diff
@@ -11,7 +11,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             name='B3',
             fields=[
-                ('id', models.AutoField(verbose_name='ID', serialize=False, primary_key=True, auto_created=True)),
+                ('id', models.BigAutoField(verbose_name='ID', serialize=False, primary_key=True, auto_created=True)),
```
Comment (auvipy) [inline tests/migrations/migrations_test_apps/lookuperror_b/migrations/0003_b3.py]: remove n regen

File: tests/migrations/migrations_test_apps/lookuperror_c/migrations/0002_c2.py [hunk 30]
```diff
@@ -12,7 +12,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             name='C2',
             fields=[
-                ('id', models.AutoField(auto_created=True, verbose_name='ID', primary_key=True, serialize=False)),
+                ('id', models.BigAutoField(auto_created=True, verbose_name='ID', primary_key=True, serialize=False)),
```
Comment (auvipy) [inline tests/migrations/migrations_test_apps/lookuperror_c/migrations/0002_c2.py]: remove

**Comment by kenneth-reitz** [comment]
I need someone to take this over, yes.

**Comment by auvipy** [comment]
> I think the main issue here is that if we were to merge this as-is it would require the generation of migrations for all models without an implicit primary key which is likely most of the models out there. If this is not detected right now it's possibly due to the `auto_created` nature of these fields that the current auto-detector might be ignoring.
> 
> Also applying these migrations would be quite costy as it would require altering all the tables backing up these models and all the tables referring to these models. In most projects that means rebuilding all of the tables managed by the ORM.
> 
> If we are to move forward with with this we need a better plan. A solution could be to go through a deprecation period where a warning is raised if a no explicit primary is defined that mentions that this models's primary key will be changed from a `AutoField` to `BigAutoField` in Django 3.x. That would allow developers to explicitly switch to `BigAutoField` if required.
> 
> By the way I don't think Basecamp is a good representation of the average Django project out there. It has been operating since 2004 on moderately large scale and they only just hit this issue. Don't get me wrong, I know that there are large Django projects out there that are likely to hit this boundary. I just feel like this isn't the case for the vast majority of them and that defaulting to `BigAutoField` might be overkill. Large projects should already have a form of DBAs in place to prevent these events from happening in the first place.

I do agree with the reasoning.

**Comment by matthiask** [comment]
> By the way I don't think Basecamp is a good representation of the average Django project out there. It has been operating since 2004 on moderately large scale and they only just hit this issue. Don't get me wrong, I know that there are large Django projects out there that are likely to hit this boundary. I just feel like this isn't the case for the vast majority of them and that defaulting to `BigAutoField` might be overkill. Large projects should already have a form of DBAs in place to prevent these events from happening in the first place.

Basecamp definitely isn't a good representation, BUT the error happened in Basecamp 3 which was only publicly available three years at the time I wrote the comment above. (https://signalvnoise.com/posts/3968-launch-basecamp-3) So it didn't take them 14 years to hit this issue, only 3 years as I understand it.

**Comment by pope1ni** [comment]
> I think the main issue here is that if we were to merge this as-is it would require the generation of migrations for all models without an implicit primary key which is likely most of the models out there. If this is not detected right now it's possibly due to the `auto_created` nature of these fields that the current auto-detector might be ignoring.
> 
> Also applying these migrations would be quite costy as it would require altering all the tables backing up these models and all the tables referring to these models. In most projects that means rebuilding all of the tables managed by the ORM.
> 
> If we are to move forward with with this we need a better plan. A solution could be to go through a deprecation period where a warning is raised if a no explicit primary is defined that mentions that this models's primary key will be changed from a `AutoField` to `BigAutoField` in Django 3.x. That would allow developers to explicitly switch to `BigAutoField` if required.

Foisting a change like this on all existing projects feels a bit unfair. This would create a lot of work for some of us with large databases - either where there are a large number of tables to update (or migrations/models to fiddle, or where there are a large number (billions) of rows to update (in a single partitioned table).

A possible solution would be to have a setting such as ``DEFAULT_AUTOFIELD`` which is set to ``'AutoField'`` by default. This could then be overridden to ``'BigAutoField'`` in the project template if we want to force new projects to use ``BigAutoField``. Anyone who wants to opt-in to the change and go through the pain and the effort of updating their project can do so.

I imagine this approach will reduce the size of this PR which currently stands as 113 files changed and we wouldn't need to introduce a migration questioner either as explicitly changing the setting could be taken as consent when generating migrations.

I also agree with @charettes:

> This should be kept as-is. We use the Small wording for small ints and omit the prefix for 32-bit integers. e.g. SmallIntegerField, IntegerField, and BigIntegerField.

This current approach feels likely to add more complexity and bugs. There are a number of efforts ongoing to try and simplify how auto fields work to allow for future improvements, e.g. non-integer auto fields, etc., and this feels like it would make that harder.

Please can there be a ticket created for this in Trac?

**Comment by felixxm** [comment]
Closing due to lack of consensus.

Thanks Kenneth and Andrew for this patch and all folks involved with discussion and reviews.

We could reconsider this patch but that would require a discussion that reached a consensus on [the mailing list](https://groups.google.com/forum/?utm_source=digest&utm_medium=email#!forum/django-developers/topics).
